### ISSUE TITLE
Telemetry event [:oban, :engine, :discard_job, :stop] is not fired.

### ISSUE THREAD

**Issue by blisscs** [sorentwo/oban#591]
### Environment

- Oban Version `2.10.1`
- PostgreSQL Version `13.5`
- Elixir & Erlang/OTP Versions (`elixir --version`)  `1.12.1`

### Current Behavior

telemetry event [:oban, :engine, :discard_job] is not fired.

My code for trying telemetry with oban looks something like below ->

```elixir
...
  :ok =
      :telemetry.attach_many(
        "oban-telemetry-handler",
        [
          [:oban, :engine, :complete_job, :stop],
          [:oban, :engine, :cancel_job, :stop],
          [:oban, :engine, :discard_job, :stop]
        ],
        &__MODULE__.handle_event/4,
        nil
      )
...
      
  def handle_event([:oban, :engine, :complete_job, :stop], _, metadata, _) do
      IO.inspect(metadata, label: "complete")
  end

  def handle_event([:oban, :engine, _, :stop], _, metadata, _) do
    IO.inspect(metadata, label: "cancel_or_fail")
  end
```

I have tested with `completed_job` and `cancel_job`, they work as expected but for `discard_job` it is not working.

### Expected Behavior

telemetry event for `discard_job` should be fired as well.

**Comment by sorentwo** [comment]
The `:discard_job` event wrapper is identical to `:complete_job` and `:cancel_job`. To verify events were fired I added a handler to the [executing test](https://github.com/sorentwo/oban/blob/main/test/integration/executing_test.exs#L11).

Here's the diff:

```diff
+      :telemetry.attach_many(
+        "oban-telemetry-handler",
+        [
+          [:oban, :engine, :complete_job, :stop],
+          [:oban, :engine, :discard_job, :stop]
+        ],
+        &__MODULE__.handle_event/4,
+        nil
+      )
+
       {:ok, name: name}
     end

+    def handle_event(event, _measure, _meta, nil) do
+      IO.inspect(event)
+    end
```

Running the tests spits out 100 or so lines like this:

```
[:oban, :engine, :discard_job, :stop]
[:oban, :engine, :complete_job, :stop]
```

Perhaps you're encountering an error or nothing is getting discarded the way you expect? Either way, there isn't anything wrong with the event emission itself 🙂

**Comment by blisscs** [comment]
@sorentwo Please see the video from the link http://gofile.me/5paKw/ZtdPYHBTs to show that the :discard_job event was not fired.

I have set up the project from the scratch and then try to inspect the event discard, cancel, and complete events.
In the video, you will see that the inspect was fired for completed and canceled ones.
And there were exceptions that happened, the job get discarded by forcing the retry and we can see from the state in the `db`.

The source code for the project I setup can be found here -> https://github.com/blisscs/oban_telemetry

**Comment by sorentwo** [comment]
@blisscs Thank you for the detailed explanation and the example. The missing `discard_job` calls were specific to exhausted retries, not manual job discards.

**Comment by blisscs** [comment]
@sorentwo Is there any recommended workarounds or shortcuts to filter and catch for :completed, :discarded, and :cancled jobs :telemetry events.

**Comment by sorentwo** [comment]
@blisscs The engine events are for span instrumentation, mostly. I recommend using `[:oban, :job, _]` events and matching on the `:state` value (which will be `:completed`, `:discarded`, etc).

**Comment by sorentwo** [comment]
Also, if it wasn't clear, the commit that closed this (cf93b14) addresses the issue you reported.

**Comment by blisscs** [comment]
Thank you very much, I just saw the reference to the commit. 

:+1:
